feat: Add new kv and fdb libs

lgsilva3087 opened the pull request:
This commit adds a basic Key-Value Engine interface based on
transactions (kv library) and a concrete implementation with
FoundationDB as backend (fdb library).

High-level applications, like MDSs, will interact with abstract
transactions, the implementation details will be hidden from them,
enabling easy switching between concrete backends.

The simple sfs_test_fdb application was modified to use the new
libraries and the new KV engine concept. It will be gradually converted
to unit and integration tests soon.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat: Add new kv and fdb libs

This commit adds a basic Key-Value Engine interface based on
transactions (kv library) and a concrete implementation with
FoundationDB as backend (fdb library).

High-level applications, like MDSs, will interact with abstract
transactions, the implementation details will be hidden from them,
enabling easy switching between concrete backends.

The simple sfs_test_fdb application was modified to use the new
libraries and the new KV engine concept. It will be gradually converted
to unit and integration tests soon.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(fdb): Create the initial unit tests

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 commented on utils/sfs_test_fdb.cc:
```diff
@@ -0,0 +1,172 @@
+	std::string startKey = "key_2";
+	std::string endKey = "key_9";
+	kv::KeySelector startSelector(kv::Key(startKey.begin(), startKey.end()), true, 0);
+	kv::KeySelector endSelector(kv::Key(endKey.begin(), endKey.end()), true, 0);
```
Fixed.

lgsilva3087 commented on tests/tools/saunafs.sh:
```diff
@@ -131,6 +134,16 @@ setup_local_empty_saunafs() {
 	done
 }
 
+init_metadata_backend() {
```
I will do it if more backends are added later.

lgsilva3087 commented on tests/tools/saunafs.sh:
```diff
@@ -131,6 +134,16 @@ setup_local_empty_saunafs() {
+			export USE_FDB
+			start_fdb_cluster
+			;;
+	esac
```
Already answered.

lgsilva3087 committed:
fix(fdb): Improve exception handling in FDBContext

Instead of calling std::terminate, most of the functions are using
exceptions now.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 commented on src/fdb/fdb_context.cc:
```diff
@@ -0,0 +1,66 @@
+
+namespace fdb {
+
+inline void check_fdb_err(int err, std::string msg) {
+	if (err != 0) {
+		safs::log_err("FoundationDB error: {}. {}", DB::errorMsg(err), msg);
+		std::terminate();
+	}
```
Applied.

lgsilva3087 commented on src/fdb/fdb_context.cc:
```diff
@@ -0,0 +1,66 @@
+	check_fdb_op(DB::setupNetwork(), "Failed to setup network thread");
+
+	networkThread_ = std::thread([] {
+		pthread_setname_np(pthread_self(), "fdb_network");
+		DB::runNetwork();
+	});
```
Applied.

lgsilva3087 commented on src/fdb/fdb_context.cc:
```diff
@@ -0,0 +1,66 @@
+	if (networkThread_.joinable()) {
+		check_fdb_op(DB::stopNetwork(), "Failed to stop network thread");
+		networkThread_.join();
+	}
```
It is hard to estimate a correct time in this early stage. I will keep it like this until it gets more mature.

lgsilva3087 commented on src/fdb/fdb.cc:
```diff
@@ -0,0 +1,207 @@
+	} else {
+		// TODO(GUILLEX): Fix the weird cast
+		safs::log_info("Transaction::get: key not found: {}",
+		               std::string(reinterpret_cast<const char *>(key.data()), key.size()));
```
Applied.

lgsilva3087 commented on src/fdb/fdb_kv_engine.h:
```diff
@@ -0,0 +1,67 @@
+	/// Creates a new transaction for reading and writing.
+	/// @return A pointer to the created transaction.
+	std::unique_ptr<kv::IReadWriteTransaction> createReadWriteTransaction() override {
+		fdb::Transaction auxTr(db_.get());
+
+		if (!auxTr) {
+			safs::log_err("FDBKVEngine::createReadWriteTransaction: Failed to create transaction");
```
Let's keep it simple for now. The caller will check for nullability.

rolysr reviewed (approved):
Very good step! @lgsilva3087

ralcolea commented on src/fdb/fdb.cc:
```diff
@@ -0,0 +1,233 @@
+				    FDB_KEYSEL_FIRST_GREATER_THAN(end.getKey().data(), end.getKey().size()), limit,
+				    kBytesLimit, streamingMode, iteration, static_cast<fdb_bool_t>(snapshot),
+				    static_cast<fdb_bool_t>(reverse));
+			} else {
```
Consider removing the `else`.

lgsilva3087 replied:
Applied.

ralcolea commented on src/fdb/fdb.cc:
```diff
@@ -0,0 +1,233 @@
+}
+
+kv::GetRangeResult Transaction::getRange(
+    const kv::KeySelector &begin, const kv::KeySelector &end, size_t limit, int iteration /* = 0 */,
```
Consider changing the `limit` type to `int` to prevent conversion type in `fdb_transaction_get_range()` function.

lgsilva3087 replied:
Applied.

ralcolea commented on src/fdb/fdb.cc:
```diff
@@ -0,0 +1,233 @@
+				    FDB_KEYSEL_FIRST_GREATER_THAN(end.getKey().data(), end.getKey().size()), limit,
+				    kBytesLimit, streamingMode, iteration, static_cast<fdb_bool_t>(snapshot),
+				    static_cast<fdb_bool_t>(reverse));
+			} else {
```
Consider removing the `else`.

lgsilva3087 replied:
Applied.

ralcolea commented on utils/sfs_test_fdb.cc:
```diff
@@ -16,125 +16,157 @@
-	if (networkThread.joinable()) {
-		checkError(fdb_stop_network());
+	// From key_2 (inclusive) to key_9 (always exclusive in FDB)
+	constexpr size_t kExpectedCount = 7;
```
`kExpectedCount` should be 8 if both ends of the range are inclusive.

lgsilva3087 replied:
Applied. Nice catches.

ralcolea reviewed (approved):
Great job @lgsilva3087!!! :+1: :fire: :muscle:

lgsilva3087 committed:
chore(fdb): Apply code review suggestions

- Use int in getRange, which matches the FDB API.
- Remove redundant else statements.
- Fix sfs_test_fdb expected range count.

Signed-off-by: guillex <guillex@leil.io>
